lib/sync/api-syncRecords.js

```diff
@@ -0,0 +1,230 @@
+ * To avoid that, we will remove those changes from both the client and the localDataset so that the client will keep them for now.
+ * @param {any} clientRecordds
+ * @param {any} localDatasetClient
+ * @param {any} pendingChanges
+ */
+function removePendingChanges(clientRecords, localDatasetClient, pendingChanges) {
+  _.each(pendingChanges, function(pendingChange, uid){
+    if (clientRecords[uid]) {
+      delete clientRecords[uid];
+    }
+    if (localDatasetClient[uid]) {
+      delete localDatasetClient[uid];
+    }
+  });
+}
+
+/**
+ * Find out the differences between the clientRecords and the serverRecords
+ * @param {String} datasetId the id of the dataset
+ * @param {Object} clientRecords client records. The key should be the uid of each record, and the value should be the hash of each record
+ * @param {Object} serverRecords server records. The key should be the uid of each record, and the value should contain the `data` and the `hash` of each record
+ * @returns the records that should be created/updated/deleted from the client
+ */
+function computeDelta(datasetId, clientRecords, serverRecords) {
+  var creates = {};
+  var updates = {};
+  var deletes = {};
+  _.each(serverRecords, function(serverRecord, serverRecordUid){
+    var serverRecHash = serverRecord.hash;
+    //record is in both client and server
+    if (clientRecords[serverRecordUid]) {
+      //hash value doesn't match, needs update
+      if (clientRecords[serverRecordUid] !== serverRecHash) {
+        syncUtil.doLog(datasetId, 'verbose', 'Updating client record ' + serverRecordUid + ' client hash=' + clientRecords[serverRecordUid]);
+        updates[serverRecordUid] = serverRecord;
+      }
+    } else {
+      //record is not in the client, needs create
+      syncUtil.doLog(datasetId, 'verbose', 'Creating client record ' + serverRecordUid);
+      creates[serverRecordUid] = serverRecord;
+    }
+  });
+
+  _.each(clientRecords, function(clientRecordHash, clientRecordUid){
+    if (!serverRecords[clientRecordUid]) {
+      //the record is in the client but not in the server, need delete
+      syncUtil.doLog(datasetId, 'verbose', 'Deleting client record ' + clientRecordUid);
+      deletes[clientRecordUid] = {};
+    }
+  });
+
+  return {
+    create: creates,
+    update: updates,
+    delete: deletes
+  };
+}
+
+/**
+ * Sync the data records for the given client
+ * @param {String} datasetId 
+ * @param {Object} params the request body, it normally contain those fields:
+ * @param {Object} params.query_params the query parameter for the dataset from the client
+ * @param {Object} params.meta_data the meta data for the dataset from the client
+ * @param {Object} params._fh an object added by the client sdk. it should have the `cuid` of the client
+ * @param {Object} params.clientRecs the client records
+ * @param {Function} cb
+ */
+function syncRecoreds(datasetId, params, cb) {
```

Comment:
I think this is a typo, `syncRecoreds` should be `syncRecords`, same in the exports